dex/elrond_dex_pair/src/lib.rs

```diff
@@ -475,14 +495,33 @@ pub trait Pair:
         let old_k = self.calculate_k_for_reserves();
 
         let mut reserve_token_out = self.pair_reserve(&token_out).get();
+        let mut virtual_reserve_token_out = self.pair_virtual_reserve(&token_out).get();
         require!(
             reserve_token_out > amount_out_min,
             "Insufficient reserve for token out"
         );
+        require!(
+            virtual_reserve_token_out > amount_out_min,
+            "Insufficient virtual reserve for token out"
+        );
 
         let mut reserve_token_in = self.pair_reserve(&token_in).get();
-        let amount_out_optimal =
-            self.get_amount_out(&amount_in, &reserve_token_in, &reserve_token_out);
+        let mut virtual_reserve_token_in = self.pair_reserve(&token_in).get();
+
+        let (reserve_token_in_to_use_for_price, reserve_token_out_to_use_for_price) = if self
+            .local_and_virtual_price_differ_too_much()
```

Comment:
This price function have the following check
`
        first_token_price_local > first_token_price_virtual_min
            && first_token_price_local < first_token_price_virtual_max`
`
So if the local price is between the boundaries for virtual price threshold it means that you want to do the swap with the virtual reserves? If this is the case, then you might need to negate this function result

Reply:
Need to negate local_and_virtual_price_differ_too_much indeed